### PR TITLE
Add support for Ruby 3.3

### PR DESCRIPTION
**Pull request by adfoster-r7**
Adds support for Ruby 3.3

## Verification

Ensure that CI passes

### CODE REVIEW

File: metasploit-framework.gemspec [hunk 1]
```diff
@@ -83,8 +83,7 @@ Gem::Specification.new do |spec|
   # NTLM authentication
   spec.add_runtime_dependency 'rubyntlm'
   # Needed by anemone crawler
-  # Locked until build env can handle newer version due to native compile issue in 1.15.x
-  spec.add_runtime_dependency 'nokogiri', '~> 1.14.0'
+  spec.add_runtime_dependency 'nokogiri'
```
Comment (adfoster-r7) [inline metasploit-framework.gemspec:86]: Nokogiri v16 is required for Ruby 3.3 support

But it looks like this was pinned previously by https://github.com/rapid7/metasploit-framework/pull/18103 when there were issues with compiling the nightly omnibus installers with older. Will require manually rebuilding ommius to see if the libxml bump in nokogiri fixes this or introduces new complexities

File: modules/payloads/singles/linux/mipsbe/exec.rb [hunk 2]
```diff
@@ -70,7 +72,9 @@ def generate(_opts = {})
     shellcode = shellcode + command_string + "\x00"
 
     # we need to align our shellcode to 4 bytes
-    (shellcode = shellcode + "\x00") while shellcode.length%4 != 0
```
Comment (adfoster-r7) [inline modules/payloads/singles/linux/mipsbe/exec.rb:73]: Bit of a weird behaviour change here between ruby versions:

```ruby
shellcode = ""

(shellcode = shellcode + "\x00") while shellcode.bytesize%4 != 0

puts RUBY_DESCRIPTION
puts shellcode.bytesize
```

Ruby 3.0:

```
$ ruby foo.rb       
ruby 3.0.2p107 (2021-07-07 revision 0db68f0233) [x86_64-darwin20]
0
```

Ruby 3.3:

```
$ ruby foo.rb
ruby 3.3.0 (2023-12-25 revision 5124f9ac75) [x86_64-darwin22]
4
```

File: modules/payloads/singles/linux/mipsbe/exec.rb [hunk 3]
```diff
@@ -70,7 +72,9 @@ def generate(_opts = {})
...
+    while shellcode.bytesize%4 != 0
+      (shellcode = shellcode + "\x00")
```
Comment (jvoisin) [inline modules/payloads/singles/linux/mipsbe/exec.rb:76]:
```suggestion
      shellcode += "\x00"
```

I don't think the parenthesis are necessary, and the short version of the *append* operator can be used :)

File: modules/payloads/singles/linux/mipsbe/exec.rb (hunk 2 above)
Comment (adfoster-r7) [reply modules/payloads/singles/linux/mipsbe/exec.rb:73]: Looks like this is fixed in the head of Ruby now; Will leave it as-is

**Comment by adfoster-r7** [comment]
This is blocked until Nokogiri 1.16.1 is released with this GCC fix for omnibus installer https://github.com/sparklemotion/nokogiri/pull/3090

**Pull request closed without merging**